0000-error-values-proposal.rst

```diff
@@ -0,0 +1,282 @@
+
+**Note**: the textual rendering of the errrors, warnings and suggestions
+should remain identical, this proposal really is about GHC's
+representation of errors.
+
+Proposed Change Specification
+-----------------------------
+
+The current representation of errors and warnings in GHC is based on the
+following data types.::
+
+    data ErrMsg = ErrMsg {
+            errMsgSpan        :: SrcSpan,
+            errMsgContext     :: PrintUnqualified,
+            errMsgDoc         :: ErrDoc,
+            errMsgShortString :: String,
+            errMsgSeverity    :: Severity,
+            errMsgReason      :: WarnReason
+            }
+
+    data ErrDoc = ErrDoc {
+            errDocImportant     :: [MsgDoc],
+            errDocContext       :: [MsgDoc],
+            errDocSupplementary :: [MsgDoc]
+            }
+
+    type MsgDoc = SDoc
+
+    -- for completeness: warnings are represented with the same type
+    -- as errors.
+    type WarnMsg = ErrMsg
+
+where ``SDoc`` is a type from GHC's pretty printing infrastructure that
+represents configurable textual documents.
+
+GHC then maintains a bag of ``ErrMsg`` and a bag of ``WarnMsg`` as
+compilation proceeds and reports them when appropriate.::
+
+    data TcLclEnv = TcLclEnv
+      { ...
+      , tcl_errs :: TcRef Messages     -- Place to accumulate errors
+      , ...
+      }
+
+    type Messages        = (WarningMessages, ErrorMessages)
+    type WarningMessages = Bag WarnMsg
+    type ErrorMessages   = Bag ErrMsg
+
+We propose to replace ``ErrDoc`` with several algebraic data types, each
+representing the different errors/warnings that might arise from a given
+GHC subsystem. For example (simplified):::
+
+    data RenamerError
+        = NotInScope OccName [Name] -- unknown name, suggestions
+	| ...
+
+    data TypecheckerError
+        = OccursCheck Type Type
+	| ...
+
+    ...
+
+We could even split error types further if necessary, making it a
+slightly more elaborate/deep hierarchy. The exact shape of the said hierarchy
+has yet to be determined, as it will be best informed by staring at the
+error generation code that GHC has today. We would also provide a toplevel
+error data type, ``GHCError``, which would be a sum of the error types from all
+subsystems. This would allow us to store and more generally treat
+uniformly errors from different systems:::
+
+    data GHCError
+      = PsError ParseError
+      | RnError RenamerError
+      | TcError TypecheckerError
+      | ...
+
+One can then define helper functions such as
+``notInScope :: OccName -> Name -> GHCError`` to be able to easily construct
+error values from within, say, the guts of a typechecker function, without
+having to make the intermediate layers and wrapping visible there. We would
+create, manipulate and store ``GHCError`` values until the very last moment,
+when it is time to render the errors and report them. This would require
+implementing ``errorMessage :: GHCError -> ErrMsg``, and would be equivalent
```

Comment:
But `ErrMsg` appears to have a bunch more information than a `GHCError`, so I don't think this is exactly right.

Instead, how about

```hs
    data ErrMsg = ErrMsg {
            errMsgSpan        :: SrcSpan,
            errMsgContext     :: PrintUnqualified,
            errMsgPayload     :: GHCError,
            errMsgShortString :: String,   -- what is this? it seems to be used only in the `Show` instance; strange!
--           errMsgSeverity    :: Severity,  -- can this now be extracted from the GHCError?
--            errMsgReason      :: WarnReason -- can this now be extracted from the GHCError?
            }
```

?

Reply:
I think I mixed up two lines of thoughts:
- embed source span info and what not in `GHCError` (but then that essentially turns `GHCError` into `ErrMsg`)
- just replace the `ErrDoc` part by `GHCError` (but then some information of `ErrMsg` is directly deduced from the error value), i.e what you suggest above

Regarding `errMsgShortString`, it does look very odd. I'd almost be tempted to suggest its removal in a new ticket, if only to see who complains and why. I guess a first step would be to just remove the field, its occurence in the `Show` instance and see what code breaks, to see where we use that `Show` instance.

Regarding the two fields that might be extracted from the error value, I do think you're right. While we can't guess the span or context, we certainly can determine the severity and warn reason from the error value, as part of the conversion from `GHCError`+some extra information to `ErrMsg`.